src/storage/txn/scheduler.rs

```diff
@@ -144,107 +146,164 @@ impl Scheduler {
+                    let pr: ProcessResult = ProcessResult::Value { value: val };
+                    let cb = make_write_cb(pr, cid, self.schedch.clone());
+                    try!(self.engine.async_write(ctx, txn.modifies(), cb));
+                }
+                Command::ScanLock { max_ts, ref mut callback, .. } => {
+                    let mut locks = vec![];
+                    let mut cursor = try!(snapshot.iter_cf("lock"));
+                    cursor.seek_to_first(); // ignore result, valid() is checked below.
+                    while cursor.valid() {
+                        let mut pb = MetaLock::new();
+                        try!(pb.merge_from_bytes(cursor.value()));
+                        if pb.get_start_ts() <= max_ts {
+                            let key = Key::from_encoded(cursor.key().to_vec());
+                            let mut lock = LockInfo::new();
+                            lock.set_primary_lock(pb.take_primary_key());
+                            lock.set_lock_version(pb.get_start_ts());
+                            lock.set_key(try!(key.raw()));
+                            locks.push(lock);
+                        }
+                        cursor.next();
+                    }
+                    callback.take().unwrap()(Ok(locks));
+                }
+                Command::ResolveLock { ref ctx, start_ts, commit_ts, ref mut callback } => {
+                    let mut keys = vec![];
+                    let mut cursor = try!(snapshot.iter_cf("lock"));
+                    cursor.seek_to_first(); // ignore result, valid() is checked below.
+                    while cursor.valid() {
+                        let mut pb = MetaLock::new();
+                        try!(pb.merge_from_bytes(cursor.value()));
+                        if pb.get_start_ts() == start_ts {
+                            keys.push(Key::from_encoded(cursor.key().to_vec()));
+                        }
+                        cursor.next();
+                    }
+                    next_cmd = Some(match commit_ts {
+                        Some(ts) => {
+                            Command::Commit {
+                                ctx: ctx.clone(),
+                                keys: keys,
+                                lock_ts: start_ts,
+                                commit_ts: ts,
+                                callback: callback.take(),
+                            }
+                        }
+                        None => {
+                            Command::Rollback {
+                                ctx: ctx.clone(),
+                                keys: keys,
+                                start_ts: start_ts,
+                                callback: callback.take(),
+                            }
+                        }
+                    });
+                }
             }
         }
-
+        if let Some(cmd) = next_cmd {
```

Comment:
I see the problem. How about making 'ResolveLock' readonly? We can treat it as a readonly command but only generate a command which is not readonly.